lib/chef/data_collector/serializers/base.rb

```diff
@@ -0,0 +1,101 @@
+#
+# Author:: Adam Leff (<adamleff@chef.io)
+# Author:: Ryan Cragun (<ryan@chef.io>)
+#
+# Copyright:: Copyright 2012-2016, Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'json'
+require 'securerandom'
+
+class Chef
+  class DataCollector
+    class Serializers
+      class Base
+        def document
+          raise "#{self.class} does not implement the #document method, which should return a hash of data to send"
+        end
+
+        def message_type
+          raise "#{self.class} does not implement the #message_type method, which return a string containing the message type"
+        end
+
+        def to_json
+          document.to_json
+        end
+
+        def chef_server_fqdn
+          return URI(Chef::Config[:chef_server_url]).host unless Chef::Config[:chef_server_url].nil?
+          return "localhost" unless defined?(run_status)
+
+          run_status.node["fqdn"] 
+        end
+
+        def organization
+          solo_run? ? data_collector_organization : chef_server_organization
+        end
+
+        def data_collector_organization
+          Chef::Config[:data_collector_organization] || "chef_solo"
+        end
+
+        def chef_server_organization
+          return nil unless Chef::Config[:chef_server_url]
+
+          Chef::Config[:chef_server_url].match(%r(/organizations/(\w+))).nil? ? "unknown_organization" : $1
+        end
+
+        def collector_source
+          solo_run? ? "chef_solo" : "chef_client"
+        end
+
+        def solo_run?
+          Chef::Config[:solo] || Chef::Config[:local_mode]
+        end
+
+        def node_uuid
```

Comment:
Would others find this helpful? To date, nothing else seemed to require the need to uniquely identify nodes like this, but because of how a Data Collector server could receive messages from a variety of different sources where node name + org isn't unique enough, I added this here and wanted to limit the blast radius. :)